joomla/joomla-cms
# [4.0] Select image/audio/video/document file from Media Manager

Pull Request for Issue https://github.com/joomla/joomla-cms/discussions/32642 and https://www.facebook.com/groups/joomlanospam/posts/10157777796480997/?__cft__[0]=AZXC8G44TJLPO0QeEbytZwSVaZX2YpNFHjOUcx1xUqyJjBUZGqgh763LWT-Lx2xKyvJfIm-LQO6wUquCrqneOLqSAlx3AOpRDAxnMrliJHTTK1AIkT4Mh-BVfJCJjMVFEyDYelefuWE8UMA3X6_-XUFRasmSPT8VNhUVlOln0Ujj4LCrwnBYH_QLfColkn2QEkk&__tn__=%2CO%2CP-R

Apparently, the Joomla community finds it totally fine to use software that is **based on a javascript that's 2 major versions behind**. It's not fine for me thus the PR...

### Summary of Changes
- Reworked the media select Javascript
- Changed the one field for all the supported extensions to multiple fileds in Media Manager (this needs to be communicated as it's a B/C break, and also it needs updates to the sql files)
- Reworked/added file types in the vue App (the files need to enable the preview)

Please extract this 
[Archive.zip](https://github.com/joomla/joomla-cms/files/6720757/Archive.zip)
 to the images directory, or use your own mp3/mp4/pdf files

### Testing Instructions
- Apply the PR or downlad the installable from github

- If you're applying the PR then go to the Media manager options and 
- Make sure the legal mime types are: `image/jpeg,image/gif,image/png,image/bmp,audio/mp3,video/mp4,application/msword,application/excel,application/pdf,application/powerpoint,text/plain,application/x-zip`
- the video extensions are `mp4`
- the audio extensions are `mp3`
- the document extensions are `odg,odp,ods,odt,pdf,png,ppt,txt,xcf,xls,csv`
- and finally the images extensions are: `bmp,gif,jpg,jpeg,png,ico` **This is IMPORTANT**
- 
- Go to tinyMCE and:

- Select the image from the CMS Content dropdown
- Then select the video file, click on the controls and embed checkboxes, change the width and height and then click select image at the bottom of the modal
- Press couple returns type `Video Link` and select the text you just wrote
- Then select the video file, without selecting the controls and embed checkboxes (width and height is irrelevant here) and then click select image at the bottom of the modal
- Check the produced HTML by clicking the toggle button under the editor

- Select the image from the CMS Content dropdown
- Then select the audio file, click on the controls and embed checkboxes and then click select image at the bottom of the modal
- Press couple returns type `Audio Link` and select the text you just wrote
- Then select the audio file, without selecting the controls and embed checkboxes and then click select image at the bottom of the modal
- Check the produced HTML by clicking the toggle button under the editor

- Select the image from the CMS Content dropdown
- Then select the PDF file, click on the embed checkboxe  change the width and height and then click select image at the bottom of the modal
- Press couple returns type `PDF Link` and select the text you just wrote
- Then select the PDF file, without selecting the embed checkboxe and then click select image at the bottom of the modal
- Check the produced HTML by clicking the toggle button under the editor

- your final html should look like:
```html
<p><video controls="controls" width="800" height="600">
<source src="/images/flower.mp4" type="video/mp4" /></video></p>
<p> </p>
<p><a href="images/flower.mp4" download="">Video Link</a></p>
<p> </p>
<p><audio src="images/t-rex-roar.mp3" controls="controls"></audio></p>
<p> </p>
<p><a href="images/t-rex-roar.mp3" download="">Audio Link</a></p>
<p> </p>
<p><object data="images/core_xy_exploded_infographic1.pdf" type="application/pdf" width="800" height="600">
You don't have a pdf plugin, but you can <a download="" href="/images/core_xy_exploded_infographic1.pdf"> download the pdf file.</a></object></p>
<p> </p>
<p><a href="images/core_xy_exploded_infographic1.pdf" download="">PDF Link</a></p>
```

- Check that the insertion of an image is still ok
- Check that the media FIELDS (intro/fulltext images) are still functional
- Finally save the article and visit it in the front end. ~That's it.~ 

- Edit the definition of the intro image https://github.com/joomla/joomla-cms/blob/7f5dd8ee5f6a690a88f05926bef6cb8456055bb8/administrator/components/com_content/forms/article.xml#L735-L739 by adding `types="images,audios"`
- Test the field, you should be able now to select images and audio files, also try to upload an unsupported file (eg video and pdf)
- Repeat with different combinations `types="images,audios,videos,documents"`


### Actual result BEFORE applying this Pull Request

Impossible to select/embed anything other than an image file


### Expected result AFTER applying this Pull Request

It's ok to select/embed audio, video, document files
<img width="865" alt="Screenshot 2021-06-26 at 23 04 08" src="https://user-images.githubusercontent.com/3889375/123526089-da132d00-d6d5-11eb-9451-3465805d6b89.png">
<img width="1133" alt="Screenshot 2021-06-26 at 23 04 38" src="https://user-images.githubusercontent.com/3889375/123526090-db445a00-d6d5-11eb-98ce-eb875a5857c7.png">
<img width="721" alt="Screenshot 2021-06-26 at 23 04 28" src="https://user-images.githubusercontent.com/3889375/123526091-dbdcf080-d6d5-11eb-86d3-4fcf685a427e.png">
<img width="857" alt="Screenshot 2021-06-26 at 23 04 18" src="https://user-images.githubusercontent.com/3889375/123526094-dbdcf080-d6d5-11eb-9eea-79cf447976d4.png">

Front end:
![j4 local_index php_teststststtsts](https://user-images.githubusercontent.com/3889375/123526488-89510380-d6d8-11eb-80fe-a2dbe85d6b4a.png)

### Documentation Changes Required

There are few todos that I skipped here:
- sql files
- remove renamed files
- rename the xtd button image to media as the image is confusing now

@wilsonge @laoneo any feedback?
